package/kernel/mac80211/files/lib/wifi/mac80211.sh

```diff
@@ -81,13 +85,39 @@ detect_mac80211() {
 		htmode=""
 		ht_capab=""
 
-		iw phy "$dev" info | grep -q 'Capabilities:' && htmode=HT20
+		iw_phy_info=$( iw phy "${dev}" info )
+
+		echo "${iw_phy_info}" | fgrep -q 'Capabilities:' && htmode=HT20
+
+		if echo "${iw_phy_info}" | egrep -q '(49|5[0-9])[0-9][0-9] MHz' ; then
 
-		iw phy "$dev" info | grep -q '5180 MHz' && {
 			mode_band="a"
-			channel="36"
-			iw phy "$dev" info | grep -q 'VHT Capabilities' && htmode="VHT80"
-		}
+
+			channel=$( echo "${iw_phy_info}" | \
+				fgrep -v disabled | \
+				sed -Ene 's/^.+ (49|5[0-9])[0-9][0-9] MHz \[([1-9][0-9]{1,2})\].*$/\2/p' | \
+				head -n 1 )
+
+			if [ -z "${channel}" ] ; then
+				logger -p warn -t detect_mac80211 \
+					"No enabled channel found in `iw phy ${dev} info`"
+			fi
+
+			if echo "${iw_phy_info}" | fgrep -q 'VHT Capabilities'
+			then
+				case "$channel" in
```

Comment:
From afar this looks like you just want to fix something for your specific device in just the US?
But is this a problem? From what I know people just do set a channel when they are setting up their
wireless security key.

Reply:
The impetus here is primarily that the config generated isn't "operational" for the radio that is limited to ch.100-165. In LuCI, you can press "Enable" and nothing happens. I agree that it is likely that the user will select a channel (though far from certain, given the number of "Why doesn't 'auto' do anything automatically?" posts seen on the forum).

As "5180" appears as a string in the output of `iw phy phy0 info` with `ieee80211-freq-limit = <5490000 5835000>;` applied, the `master` implementation should at least pick up that it should be configured as a 5 GHz radio. 

I have confirmed that the `master` version identifies and provides config for all three radios, even with the freq. limits in place.

I have confirmed that LuCI shows the limited set of channels in is drop-downs.

Things get even more interesting with the -AH cal files and JP as a country code, where it looks like there aren't any available channels for the second radio (from poking at it, as I haven't found a tool to decode these pre-cal/board combinations, only the older all-in-ART ones).

I did try `channel=acs_survey` and it seemed to crash the wireless, at least when used in the context of `option channel 'acs_survey'` in `/etc/config/wireless` (assuming that the parameter value would be copied verbatim to the hostapd config). I did not pursue it further.

While not the best user experience, if there are no objections, at this time I'd like to put 802.11 config generation off into a separate exploration. The more I look at it, the more challenging it becomes to do in a way that is significantly better for the less-experienced user.

Reply:
> The impetus here is primarily that the config generated isn't "operational" for the radio that is limited to ch.100-165. In LuCI, you can press "Enable" and nothing happens. I agree that it is likely that the user will select a channel (though far from certain, given the number of "Why doesn't 'auto' do anything automatically?" posts seen on the forum).

This is all well and true. But from experience with DFS and regdomain, I think the solution here is to make "auto" work. And since your patch touches on this very area, I think a closer look at the userspace wpad (hostapd / wpa_supplicant) and the kernel (mt76) is duly warranted here. Because what else can you do that will work for the default?
You already touched on the regdb.txt issue with JP and AH. Furthermore channels can be dynamically disabled because of an operating radar nearby so that device has to either move to a different channel or cease operation completely. Really, I think the hostapd/wpad is the only place where a sensible channel and operation mode(HT20/40, VHT40/80/80+80/160) can be performed.

(Of course, if you don't want to deal with that side of issues as well, you can just do what other did and just drop the patch and it will be the end of that).

> As "5180" appears as a string in the output of iw phy phy0 info with ieee80211-freq-limit = <5490000 5835000>; applied, the master implementation should at least pick up that it should be configured as a 5 GHz radio.

I think this is something that @ecsv might be interested in as well. The A62's 5GHz wifis also have set limits for their operational range due to in-line band filters. So I wonder how the OpenMesh deals with the problem or if they are just providing a customized /etc/config/wireless in their factory image and the wifi detection is skipped.

> I did try channel=acs_survey and it seemed to crash the wireless, at least when used in the context of option channel 'acs_survey' in /etc/config/wireless (assuming that the parameter value would be copied verbatim to the hostapd config). I did not pursue it further.

Yes, ```channel=acs_survey``` is the hostapd.conf option. From what I know the /e(tc)/c(config)/wireless option is ```auto```. But I think it's also possible to omit the channel option in /e/c/wireless. 

> While not the best user experience, if there are no objections, at this time I'd like to put 802.11 config generation off into a separate exploration. The more I look at it, the more challenging it becomes to do in a way that is significantly better for the less-experienced user.

True, I was at odds with the wifi detection code in the past as well 5f8f8a366136a07df661e31decce2458357c167a 5e35b4562fb73aaa1c75a3404fed2efb6afe03f8 . 
From what I remember the issue with the channel selection also creeped up during the discussion, but there has been no push to do anything about it.

Reply:
> I think this is something that @ecsv might be interested in as well. The A62's 5GHz wifis also have set limits for their operational range due to in-line band filters. So I wonder how the OpenMesh deals with the problem or if they are just providing a customized /etc/config/wireless in their factory image and the wifi detection is skipped.

There is no OpenMesh anymore - all part of Datto Networking now.

I am not involved in this part but they are using a replacement of netifd at the moment (and no Luci). Something which does additional checks before accepting channels for specific radios and so on. And the wireless configuration is also (partially) done using some different mechanism (don't ask me why).